performance/multi-node/rdm_mbw_mr.c

```diff
@@ -148,52 +169,117 @@ static void cq_readerr(struct fid_cq *cq, const char *cq_str)
 /*
  * fi_cq_err_entry can be cast to any CQ entry format.
  */
-static int wait_for_comp(struct fid_cq *cq, int num_completions)
+static int wait_for_comp(struct fid_cntr *cntr, struct fid_cq *cq,
+			 int num_completions)
 {
 	struct fi_cq_err_entry comp;
 	int ret;
 
-	while (num_completions > 0) {
-		ret = fi_cq_read(cq, &comp, 1);
-		if (ret > 0) {
-			num_completions--;
-		} else if (ret < 0 && ret != -FI_EAGAIN) {
-			if (ret == -FI_EAVAIL) {
-				cq_readerr(cq, "cq");
-			} else {
-				ct_print_fi_error("fi_cq_read", ret);
+	while (1) {
+		ret = 0;
+		if (test_type == CQ_TEST) {
+			if (num_completions <= 0)
+				return 0;
+
+			if (wait_type == FI_WAIT_UNSPEC)
+				ret = fi_cq_sread(cq, &comp, 1, NULL, -1);
+			else
+				ret = fi_cq_read(cq, &comp, 1);
+			if (ret > 0) {
+				num_completions--;
+				continue;
+			} else if (ret < 0 && ret != -FI_EAGAIN) {
+				if (ret == -FI_EAVAIL) {
+					cq_readerr(cq, "cq");
+				} else {
+					ct_print_fi_error("fi_cq_read", ret);
+				}
+				return ret;
 			}
-			return ret;
 		}
+
+		if (test_type == CNTR_TEST) {
+			if (fi_cntr_read(cntr) >= num_completions) {
+				return 0;
+			}
+
+			if (wait_type == FI_WAIT_UNSPEC)
+				ret = fi_cntr_wait(cntr, num_completions, -1);
+		}
+
+		if (wait_type == FI_WAIT_SET) {
+			ret = fi_wait(waitset, -1);
+		}
+
+		if (ret < 0 && (wait_type == FI_WAIT_SET ||
+				wait_type == FI_WAIT_UNSPEC)) {
+			fprintf(stdout, "Waiting timed out %d\n", ret);
```

Comment:
So it shouldn't ever timeout due to the usage of -1 to fi_wait/fi_cntr_wait, so the refactor keeps essentially the same exit conditions as the original, if wanted I could also add a timeout control variable to this to allow the user to change the default timeout for the fi_cntr_wait and fi_wait options.